cmd/query/main.go

```diff
@@ -89,9 +103,27 @@ func main() {
 			sHandler.RegisterRoutes(r)
 			portStr := ":" + strconv.Itoa(queryOpts.QueryPort)
 			recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
-			logger.Info("Starting jaeger-query HTTP server", zap.Int("port", queryOpts.QueryPort))
-			if err := http.ListenAndServe(portStr, recoveryHandler(r)); err != nil {
-				logger.Fatal("Could not launch service", zap.Error(err))
+
+			go func() {
+				// there's still a chance that a request comes in between the `ready` flag and the server's
+				// actual startup, but this would be a rare situation, as we assume the `ListenAndServe` is fast enough
+				// between allocating a port and serving a request.
+				hcChannel <- http.StatusNoContent
+
+				logger.Info("Starting jaeger-query HTTP server", zap.Int("port", queryOpts.QueryPort))
+				if err := http.ListenAndServe(portStr, recoveryHandler(r)); err != nil {
```

Comment:
+1 . Should this be outside of the go routine, before the select on the channel? There's a chance that the health check server would be marked as ready before the main server starts, but it should not be such a big issue, I think.